packages/vite/src/node/ssr/runtime/utils.ts

```diff
@@ -0,0 +1,276 @@
+import type { ImportMetaEnv } from './types'
+
+export const isWindows =
+  typeof process !== 'undefined' && process.platform === 'win32'
+
+export const decodeBase64 =
+  typeof atob !== 'undefined'
+    ? atob
+    : (str: string) => Buffer.from(str, 'base64').toString('utf-8')
+
+// currently we copy this from '../../constants' - maybe we can inline it somewhow?
+const NULL_BYTE_PLACEHOLDER = `__x00__`
+const VALID_ID_PREFIX = `/@id/`
+
+export function wrapId(id: string): string {
+  return id.startsWith(VALID_ID_PREFIX)
+    ? id
+    : VALID_ID_PREFIX + id.replace('\0', NULL_BYTE_PLACEHOLDER)
+}
+
+export function unwrapId(id: string): string {
+  return id.startsWith(VALID_ID_PREFIX)
+    ? id.slice(VALID_ID_PREFIX.length).replace(NULL_BYTE_PLACEHOLDER, '\0')
+    : id
+}
+
+const windowsSlashRE = /\\/g
+export function slash(p: string): string {
+  return p.replace(windowsSlashRE, '/')
+}
+
+export const queryRE = /\?.*$/s
+export const hashRE = /#.*$/s
+
+export function cleanUrl(url: string): string {
+  return url.replace(hashRE, '').replace(queryRE, '')
+}
```

Comment:
We have improved the perf for `cleanUrl` so it's only a single regex: https://github.com/sheremet-va/vite/blob/fea94b53e79990e1104651d9fb22d6954284b197/packages/vite/src/node/utils.ts#L298-L301